token/confidential-transfer/proof-generation/src/mint.rs

```diff
@@ -0,0 +1,88 @@
+use {
+    crate::{encryption::MintAmountCiphertext, errors::TokenProofGenerationError, try_split_u64},
+    solana_zk_sdk::{
+        encryption::{elgamal::ElGamalPubkey, pedersen::Pedersen},
+        zk_elgamal_proof_program::proof_data::{
+            BatchedGroupedCiphertext3HandlesValidityProofData, BatchedRangeProofU64Data,
+        },
+    },
+};
+
+const MINT_AMOUNT_LO_BIT_LENGTH: usize = 16;
+const MINT_AMOUNT_HI_BIT_LENGTH: usize = 32;
+/// The padding bit length in range proofs to make the bit-length power-of-2
+const RANGE_PROOF_PADDING_BIT_LENGTH: usize = 16;
+
+/// The proof data required for a confidential mint instruction
+pub struct MintProofData {
+    pub ciphertext_validity_proof_data: BatchedGroupedCiphertext3HandlesValidityProofData,
+    pub range_proof_data: BatchedRangeProofU64Data,
+}
+
+pub fn mint_split_proof_data(
+    mint_amount: u64,
```

Comment:
Sorry if this is a silly question, because I think you've described it before, but I don't remember the answer.

The mint proof looks like it just adds whatever ciphertext is provided, as long as it's well-formed. Is there a check on the resulting amount in the token account at some point? What happens if you mint `u64::MAX` on a loop into someone's account?

Reply:
Yeah this is actually a good question. There could be two issues here. One is that the destination account ciphertext could be non-decryptable because the underlying encrypted value grows too large. This is not an issue because the mint amount is restricted to 48-bits (by range proof) and gets added to the pending balance instead of the available balance. There is a pending balance credit counter (https://github.com/solana-labs/solana-program-library/blob/master/token/program-2022/src/extension/confidential_transfer/mod.rs#L105) to limit this amount and the receiver can always do `ApplyPendingBalance` to flush this out.

The second issue is that the global supply itself can surpass 2^64. If each mint amount is 2^48 max, then after 2^16 ~ 65k mints, the total supply can surpass 2^64. The ElGamal ciphertexts (and Pedersen commitments) can store up to 256-bit amounts, so the supply can technically go up to this amount (it will never be able to surpass this number because you need to mint 2^208 times, which is practically impossible). I did not add it, but if we do not want the supply to surpass 2^64, then we can add an extra range-proof in the mint proof certifying that the supply ciphertext is always a 64-bit number. What do you think?

Reply:
Hm, that's a tough one. SPL Token mints are normally limited to a max supply of 2^64, so my inclination is to force the same thing here. Otherwise, people can get in strange situations where they try to deserialize their balance as a u64 and get unexpected values. If it were a totally new protocol, we could avoid the constraint totally and go with 2^256, but I worry about downstream issues. Unless there's a compelling reason for the supply to go up to 2^256, I think we'll need that extra range proof.